Documentation/design-docs/tiered-compilation.md

```diff
@@ -0,0 +1,95 @@
+
+  **Steady-State** - If code loaded from ReadyToRun images appears hot, the runtime replaces it jitted code which is typically higher quality. At runtime the JIT is able to observe the exact dependencies that are loaded as well as CPU instruction support which allows it generate superior code. In the future it may also utilize profile guided feedback but it does not currently do so.
+
+
+
+Dependencies
+============
+
+## Tiered Compilation Depends On ##
+
+### The CodeVersionManager ###
+The CodeVersionManager assists in creating new method implementations for a given .NET method,  configuring those versions and (indirectly) producing their code, keeping track of which ones exist, and switching which implementation is currently active.
+
+There is a lot of useful background information in the [CodeVersionManager spec](./code-versioning.md) to understand general concepts around versioning of methods.
+
+### The Prestub ###
+The Prestub provides an initial version agnostic implementation of any method before code is available. After code is available Tiered Compilation continues to use it to count calls to a method before dispatching to the currently active code.
+
+### The Threadpool ###
+The threadpool provides the threads that are used to background compile better versions of a method's code. It also provides timer callbacks used for some aspects of tiered compilation policy.
+
+## Components that depend on Tiered Compilation ##
+
+### Interpreter ###
+The IL interpreter (disabled by default with both build defines and runtime switches) now uses tiered compilation to asynchronously promote from interpreted to jitted code.
+
+Design
+======
+
+## Historical context ##
+
+Tiered Compilation was prototyped in 2016, introduced into the runtime code in 2017, and offered as an opt-in Preview feature in .Net Core 2.1 RTM in 2018. This design doc was written after the fact. We had been trying to mitigate runtime startup and performance problems for nearly 20 years with various forms of pre-compilation (NGEN, ReadyToRun, MulticoreJit) but this was the first serious foray into using compilation tiers to achieve similar goals. The IL interpreter appears similar but as best I understand it was not primarily targeted at performance, but rather at portability into environments that did not allow jitting. Although the idea of tiered compilation had come up repeatedly in the past it had never gained the degree of consensus/acceptance/momentum necessary to move forward relative to other performance investments.
+
+## Goals ##
+
+1. Improve the steady state and startup performance of typical .Net Core workloads while minimizing regressions.
+2. Compliment existing precompilation techniques so that developers so that .NET can leverage the best of both options.
```

Comment:
> so that developers so that .NET can leverage the best of both options.

so that developers ~~so that .NET~~ can leverage the best of both options.